coordinator-common/src/main/java/org/apache/kafka/coordinator/common/runtime/CoordinatorRuntime.java

```
@@ -840,7 +848,11 @@ private void failCurrentBatch(Throwable t) {
             if (currentBatch != null) {
                 coordinator.revertLastWrittenOffset(currentBatch.baseOffset);
                 for (DeferredEvent event : currentBatch.deferredEvents) {
-                    event.complete(t);
+                    try {
+                        event.complete(t);
+                    } catch (Throwable e) {
+                        log.error("Event {} for {} failed to complete.", event, tp, e);
+                    }
```

Instead of putting those try..catch everywhere we call complete, I wonder if we could have a wrapper event which would catch all exceptions. We could wrap all the events with it. We need only one instance. Have you considered something like this? I am not sure if it is really better though.

---

I'm very tempted to try something like this. Or maybe wrap each event in a wrapper? Or have an abstract base class that does the catching and have the concrete classes override a different method for their completions.

---

Yeah, a wrapper could be a nice approach. Wanna give it a try?